tools/check_test_name.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python
+
+"""
+Check that test names start with `test`, and that test classes start with
+`Test`.
+"""
+from __future__ import annotations
+
+import ast
+import os
+from pathlib import Path
+import sys
+from typing import Iterator, Sequence
+
+PRAGMA = "# not a test"
+
+def _find_names(node: ast.Module) -> Iterator[str]:
+    for _node in ast.walk(node):
+        if isinstance(_node, ast.Name):
+            yield _node.id
+        elif isinstance(_node, ast.Attribute):
+            yield _node.attr
+
+
+def _is_fixture(node: ast.expr) -> bool:
+    if isinstance(node, ast.Call):
+        node = node.func
+    return (
+        isinstance(node, ast.Attribute)
+        and node.attr == "fixture"
+        and isinstance(node.value, ast.Name)
+        and node.value.id == "pytest"
+    )
+
+
+def _is_register_dtype(node):
+    return (isinstance(node, ast.Name) and node.id == "register_extension_dtype")
+
+
+def is_misnamed_test_func(
+    node: ast.expr | ast.stmt, names: Sequence[str], line: str
+) -> bool:
+    return (
+        isinstance(node, ast.FunctionDef)
+        and not node.name.startswith("test")
+        and names.count(node.name) == 0
+        and not any(_is_fixture(decorator) for decorator in node.decorator_list)
+        and PRAGMA not in line
+        and node.name
+        not in ("teardown_method", "setup_method", "teardown_class", "setup_class")
+    )
+
+
+def is_misnamed_test_class(
+    node: ast.expr | ast.stmt, names: Sequence[str], line: str
+) -> bool:
+    return (
+        isinstance(node, ast.ClassDef)
+        and not node.name.startswith("Test")
+        and names.count(node.name) == 0
+        and not any(_is_register_dtype(decorator) for decorator in node.decorator_list)
+        and PRAGMA not in line
+        # Some of the KDTreeTest use a decorator to setup tests so these are
+        # actually fine
+        and "KDTreeTest" not in [decorator.id for decorator in node.decorator_list]
```

Comment:
There is the possibility of more edge cases such as this. However I propose they are added as and when during the fixing of the miss named tests rather than trying to do a an exhaustive check now.